monkestation/code/modules/antagonists/vampire/actions/innate/recuperation.dm

```diff
@@ -0,0 +1,282 @@
+/datum/action/cooldown/vampire/recuperation
+	name = "Recuperation"
+	desc = "Greatly increase your recovery rate beyond that of any mortal. The process drains lifeforce and doesn't work in masquerade."
+	button_icon_state = "power_recup"
+	check_flags = NONE
+	toggleable = TRUE
+	starts_active = TRUE
+	works_in_masquerade = TRUE // on_life still prevents it from working, but you can toggle it if you want
+	has_custom_life_cost = TRUE
+
+	/// To avoid situations where a vampire has to wait 2 years to regrow stuff because of RNG, it uses slightly randomized accumulation instead. If this reaches 1, something regrows.
+	var/regrow_accumulation = 0
+
+	/// Increments while when we're dead and able to revive, resets otherwise. If this reaches 1, we revive.
+	var/revival_progress = 0
+
+	/// Accumulation var for brain traumas, when this reaches 1, we heal a brain trauma.
+	var/trauma_heal_progress = 0
+
+/datum/action/cooldown/vampire/recuperation/New(Target)
+	. = ..()
+	RegisterSignal(vampire, COMSIG_VAMPIRE_STAT_CHANGED_MOD, PROC_REF(on_stat_changed))
+	update_recovery_scaling(vampire.get_stat_modified(VAMPIRE_STAT_RECOVERY))
+
+/datum/action/cooldown/vampire/recuperation/Destroy()
+	. = ..()
+	UnregisterSignal(vampire, COMSIG_VAMPIRE_STAT_CHANGED_MOD)
+
+/datum/action/cooldown/vampire/recuperation/on_toggle_on()
+	RegisterSignal(owner, COMSIG_LIVING_LIFE, PROC_REF(on_life))
+
+/datum/action/cooldown/vampire/recuperation/on_toggle_off()
+	UnregisterSignal(owner, COMSIG_LIVING_LIFE)
+	regrow_accumulation = 0
+	revival_progress = 0
+
+/datum/action/cooldown/vampire/recuperation/proc/on_life(datum/source, seconds_per_tick, times_fired)
+	SIGNAL_HANDLER
+
+	if(vampire.masquerade_enabled)
+		return
+
+	var/regen_rate = vampire.regen_rate_modifier.get_value()
+	var/regen_level = vampire.get_stat(VAMPIRE_STAT_RECOVERY)
+
+	if(regen_rate <= 0) // sanity check
+		return
+
+	if(owner.stat == DEAD)
```

Comment:
No extra healing in a coffin? Vamp species does

Reply:
Planned, as stated in the HackMD.